Add stubs for Digital Credentials API

marcoscaceres opened the pull request:
#### 6e2d578efd74077877f2c5f0364d9735a6a188f1
<pre>
Add stubs for Digital Credentials API
<a href="https://bugs.webkit.org/show_bug.cgi?id=267844">https://bugs.webkit.org/show_bug.cgi?id=267844</a>
<a href="https://rdar.apple.com/problem/121408392">rdar://problem/121408392</a>

Reviewed by Andy Estes.

Implements the initial stubs based on the draft explainer at:
<a href="https://github.com/WICG/identity-credential/blob/main/digital-credentials-2-proposal.md">https://github.com/WICG/identity-credential/blob/main/digital-credentials-2-proposal.md</a>

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Modules/credentialmanagement/CredentialsContainer.cpp:
(WebCore::CredentialsContainer::requestIdentity):
* Source/WebCore/Modules/credentialmanagement/CredentialsContainer.h:
* Source/WebCore/Modules/credentialmanagement/CredentialsContainer.idl:
* Source/WebCore/Modules/credentialmanagement/DigitalCredential.cpp: Copied from Source/WebCore/Modules/credentialmanagement/CredentialsContainer.idl.
(WebCore::DigitalCredential::create):
(WebCore::DigitalCredential::DigitalCredential):
* Source/WebCore/Modules/credentialmanagement/DigitalCredential.h: Copied from Source/WebCore/Modules/credentialmanagement/CredentialsContainer.idl.
(WebCore::DigitalCredential::response const):
* Source/WebCore/Modules/credentialmanagement/DigitalCredential.idl: Copied from Source/WebCore/Modules/credentialmanagement/CredentialsContainer.idl.
* Source/WebCore/Modules/credentialmanagement/IdentityCredentialProtocol.h: Copied from Source/WebCore/Modules/credentialmanagement/CredentialsContainer.idl.
* Source/WebCore/Modules/credentialmanagement/IdentityCredentialProtocol.idl: Copied from Source/WebCore/Modules/credentialmanagement/CredentialsContainer.idl.
* Source/WebCore/Modules/credentialmanagement/IdentityRequestOptions.h: Copied from Source/WebCore/Modules/credentialmanagement/CredentialsContainer.idl.
* Source/WebCore/Modules/credentialmanagement/IdentityRequestOptions.idl: Copied from Source/WebCore/Modules/credentialmanagement/CredentialsContainer.idl.
* Source/WebCore/Modules/credentialmanagement/IdentityRequestProvider.h: Copied from Source/WebCore/Modules/credentialmanagement/CredentialsContainer.idl.
* Source/WebCore/Modules/credentialmanagement/IdentityRequestProvider.idl: Copied from Source/WebCore/Modules/credentialmanagement/CredentialsContainer.idl.
* Source/WebCore/Modules/mediastream/RTCIceTransport.cpp:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/bindings/js/WebCoreBuiltinNames.h:

Canonical link: <a href="https://commits.webkit.org/273545@main">https://commits.webkit.org/273545@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c7b663f44ddef100c52c724383a776b9d3b2b652

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/35592 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/14534 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/37730 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/38333 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/32069 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/36790 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/16919 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/11560 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/30870 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/36145 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/12280 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/31673 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/10787 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/10801 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/31811 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/39579 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/30154 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/32343 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/32148 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/36751 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/35458 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/10987 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/8872 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/34829 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/12699 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/42148 "Built successfully") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/8150 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/11496 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/8752 "Found 39 new JSC stress test failures: microbenchmarks/array-from-derived-object-func.js.bytecode-cache, microbenchmarks/array-from-derived-object-func.js.default, microbenchmarks/array-from-derived-object-func.js.dfg-eager, microbenchmarks/array-from-derived-object-func.js.dfg-eager-no-cjit-validate, microbenchmarks/array-from-derived-object-func.js.eager-jettison-no-cjit, microbenchmarks/array-from-derived-object-func.js.no-cjit-collect-continuously, microbenchmarks/array-from-derived-object-func.js.no-cjit-validate-phases, microbenchmarks/array-from-derived-object-func.js.no-llint, microbenchmarks/array-from-object-func.js.bytecode-cache, microbenchmarks/array-from-object-func.js.default ... (failure)") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/11767 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

annevk commented on Source/WTF/wtf/PlatformEnable.h:
```diff
@@ -216,6 +216,10 @@
 
+#if !defined(ENABLE_DIGITAL_CREDENTIALS)
+#define ENABLE_DIGITAL_CREDENTIALS 0
+#endif
```
Is there a need for a build-time flag?

marcoscaceres replied:
Maybe not. I was unsure. I'll get rid of it.

annevk replied:
It seems like you only removed the PlatformEnable change.

aestes commented on Source/WebCore/Modules/credentialmanagement/CredentialsContainer.h:
```diff
@@ -39,9 +40,11 @@ enum class Scope;
 namespace WebCore {
 
 class Document;
+class DigitalCredential;
```
Nit: should come before `Document`

aestes commented on Source/WebCore/Modules/credentialmanagement/DigitalCredential.h:
```diff
@@ -0,0 +1,59 @@
+#if ENABLE(DIGITAL_CREDENTIALS)
+
+#include "IDLTypes.h"
+#include <wtf/RefCounted.h>
```
For completeness you should include `wtf/RefPtr.h` here since that's needed for `m_response`.

aestes commented on Source/WebCore/Modules/credentialmanagement/IdentityRequestOptions.h:
```diff
@@ -0,0 +1,46 @@
+#include "AbortSignal.h"
+#include "IdentityRequestProvider.h"
+#include <wtf/RefCounted.h>
+#include <wtf/Vector.h>
```
Same comment about `wtf/RefPtr.h`

aestes commented on Source/WebCore/Sources.txt:
```diff
@@ -3382,6 +3383,7 @@ JSCustomEffectCallback.cpp
 JSCustomElementRegistry.cpp
 JSCustomEvent.cpp
 JSCustomStateSet.cpp
+JSDigitalCredential.cpp
```
Nit: this should be inserted in ASCII order (i.e. below the `JSDOM`s)

annevk commented on Source/WebCore/Modules/credentialmanagement/CredentialsContainer.cpp:
```diff
@@ -32,8 +32,10 @@
 #include "Document.h"
 #include "ExceptionOr.h"
+#include "IdentityRequestOptions.h"
 #include "JSDOMPromiseDeferred.h"
```
I suspect you have to include JSDigitialCredential.h here to address the build error. (Or something to that effect in one of these files.)

marcoscaceres replied:
Amazing! thank you.

marcoscaceres committed:
Add stubs for Digital Credentials API
https://bugs.webkit.org/show_bug.cgi?id=267844
rdar://problem/121408392

Reviewed by Andy Estes.

Implements the initial stubs based on the draft explainer at:
https://github.com/WICG/identity-credential/blob/main/digital-credentials-2-proposal.md

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Modules/credentialmanagement/CredentialsContainer.cpp:
(WebCore::CredentialsContainer::requestIdentity):
* Source/WebCore/Modules/credentialmanagement/CredentialsContainer.h:
* Source/WebCore/Modules/credentialmanagement/CredentialsContainer.idl:
* Source/WebCore/Modules/credentialmanagement/DigitalCredential.cpp: Copied from Source/WebCore/Modules/credentialmanagement/CredentialsContainer.idl.
(WebCore::DigitalCredential::create):
(WebCore::DigitalCredential::DigitalCredential):
* Source/WebCore/Modules/credentialmanagement/DigitalCredential.h: Copied from Source/WebCore/Modules/credentialmanagement/CredentialsContainer.idl.
(WebCore::DigitalCredential::response const):
* Source/WebCore/Modules/credentialmanagement/DigitalCredential.idl: Copied from Source/WebCore/Modules/credentialmanagement/CredentialsContainer.idl.
* Source/WebCore/Modules/credentialmanagement/IdentityCredentialProtocol.h: Copied from Source/WebCore/Modules/credentialmanagement/CredentialsContainer.idl.
* Source/WebCore/Modules/credentialmanagement/IdentityCredentialProtocol.idl: Copied from Source/WebCore/Modules/credentialmanagement/CredentialsContainer.idl.
* Source/WebCore/Modules/credentialmanagement/IdentityRequestOptions.h: Copied from Source/WebCore/Modules/credentialmanagement/CredentialsContainer.idl.
* Source/WebCore/Modules/credentialmanagement/IdentityRequestOptions.idl: Copied from Source/WebCore/Modules/credentialmanagement/CredentialsContainer.idl.
* Source/WebCore/Modules/credentialmanagement/IdentityRequestProvider.h: Copied from Source/WebCore/Modules/credentialmanagement/CredentialsContainer.idl.
* Source/WebCore/Modules/credentialmanagement/IdentityRequestProvider.idl: Copied from Source/WebCore/Modules/credentialmanagement/CredentialsContainer.idl.
* Source/WebCore/Modules/mediastream/RTCIceTransport.cpp:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/bindings/js/WebCoreBuiltinNames.h:

Canonical link: https://commits.webkit.org/273545@main
